atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionManager.java

```diff
@@ -240,19 +251,39 @@ public List<OpenTransaction> startTransactions(List<? extends PreCommitCondition
             openTransactionCounter.inc(transactions.size());
             return transactions;
         } catch (Throwable t) {
-            responses.forEach(response -> lockWatchManager.requestTransactionStateRemovalFromCache(
-                    response.startTimestampAndPartition().timestamp()));
-            timelockService.tryUnlock(responses.stream()
+            // In case of failure, we need to clean up the resources opened in startTransaction, for all transactions
+            // opened.
+            // These resources are the preCommitCondition, removing transaction from lock-watch and unlocking immutable
+            // timestamp.
+            // Note that in case we don't throw, the immutable timestamp is registered to be unlocked in the
+            // SnapshotTransaction constructor. But in case of failure, we need to manually unlock it here in case
+            // of failures though.
+
+            // N.B. using closer to run all cleanup tasks even if one cleanup throws
+            Closer closer = Closer.create();
+
+            conditions.forEach(condition -> closer.register(condition::cleanup));
```

Comment:
Note that if we throw before this line or this try block (e.g. in L218) then we'd fail to cleanup the condition. I might try-catch this in a higher block, same as L196-202.